sale_delivery_date/models/sale_order_line.py

```diff
@@ -6,124 +6,201 @@
-                self.order_id.name,
-            )
-            return
-        # If no commitment date is set, we must consider next preferred delivery
-        #  window to postpone date_planned
-
-        # Remove security lead time to ensure the delivery date (and not the
-        # date planned of the picking) will match delivery windows
-        date_planned_without_sec_lead = date_planned + timedelta(
-            days=self.order_id.company_id.security_lead
+        partner = self.order_id.partner_shipping_id
+        warehouse = self.order_id.warehouse_id
+        calendar = warehouse.calendar_id
+        date_planned = self._apply_cutoff(date_planned, partner, warehouse)
+        date_planned = self._apply_workload(date_planned, workload_days, calendar)
+        date_deadline = self._apply_delivery_window(
+            date_planned, partner, security_lead, calendar
         )
-        ops = self.order_id.partner_shipping_id
-        next_preferred_date = ops.next_delivery_window_start_datetime(
-            from_date=date_planned_without_sec_lead
+        date_planned = self._deduce_workload_and_security_lead(
+            date_deadline, partner, warehouse, workload_days, security_lead
         )
-        # Add back security lead time
-        next_preferred_date_with_sec_lead = next_preferred_date - timedelta(
-            days=self.order_id.company_id.security_lead
+        res.update({"date_deadline": date_deadline, "date_planned": date_planned})
+        return res
+
+    @api.model
+    def _deduce_workload_and_security_lead(
+        self, date_deadline, partner, warehouse, days, security_lead
+    ):
+        """Return the 'date_planned' from the 'date_deadline'.
+
+        Once we know the delivery date of the customer, we are able to
+        compute the final scheduled date of the transfer by taking into
+        account the workload, the WH calendar and the cutoff time.
+        """
+        # Remove the security lead from the date_deadline
+        # If there's a calendar, remove 1 day until we get to a working day.
+        # The reason for that is the parcel cannot be given to the carrier
+        # on a non-working day.
+        calendar = warehouse.calendar_id
+        date_planned = date_deadline - timedelta(days=security_lead)
+        if calendar:
+            next_working_day = self._next_working_day(date_planned, calendar)
+            count = 0
+            while next_working_day.date() != date_planned.date():
+                if count > FIND_WORKING_DAY_COUNT:  # To avoid infinite loop
+                    raise exceptions.UserError(
+                        _(
+                            "Unable to find a working day matching "
+                            "customer's delivery time window."
+                        )
+                    )
+                date_planned -= timedelta(days=1)
+                next_working_day = self._next_working_day(date_planned, calendar)
+                count += 1
```

Comment:
If the time is not a working time, your implementation won't work.
This should do what you want and the time must be carefully set before
```suggestion
            date_planned = calendar.plan_hours(-1, date_planned, compute_leaves=True)
```